File: Sources/AIKit/OpenRouterProvider.swift
import Foundation

// MARK: - OpenRouterProvider

/// `AIProvider` for OpenRouter (`POST https://openrouter.ai/api/v1/chat/completions`).
///
/// OpenRouter is OpenAI-compatible: same body shape, same response parser. The two
/// notable differences are (a) the model identifier is `vendor/model` instead of a
/// bare model name, and (b) OpenRouter rate-limits more aggressively unless you set
/// the `HTTP-Referer` / `X-Title` headers, which they use to attribute requests on
/// the dashboard.
public final class OpenRouterProvider: AIProvider, Sendable {

    public typealias HTTPClient = @Sendable (URLRequest) async throws -> (Data, URLResponse)

    // MARK: Stored

    private let apiKey: String
    private let endpoint: URL
    private let referer: String
    private let title: String
    private let httpClient: HTTPClient

    // MARK: Init

    public init(
        apiKey: String,
        endpoint: URL = OpenRouterProvider.defaultEndpoint,
        referer: String = "https://jarvisnote.studio",
        title: String = "Jarvis Note",
        httpClient: @escaping HTTPClient = OpenRouterProvider.defaultHTTPClient
    ) {
        self.apiKey = apiKey
        self.endpoint = endpoint
        self.referer = referer
        self.title = title
        self.httpClient = httpClient
    }

    // MARK: Defaults

    public static let defaultEndpoint = URL(string: "https://openrouter.ai/api/v1/chat/completions")!
    public static let defaultModel = "anthropic/claude-3.5-sonnet"

    public static let defaultHTTPClient: HTTPClient = { request in
        try await URLSession.shared.data(for: request)
    }

    // MARK: AIProvider

    public func chat(messages: [ChatMessage], config: AIConfig) async throws -> String {
        let request = try Self.buildRequest(
            endpoint: endpoint,
            apiKey: apiKey,
            referer: referer,
            title: title,
            messages: messages,
            config: config
        )

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await httpClient(request)
        } catch {
            throw AIError.sendFailed(message: error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIError.sendFailed(message: "Non-HTTP response from OpenRouter")
        }

        switch httpResponse.statusCode {
        case 200:
            // Same response shape as OpenAI — reuse its parser.
            return try OpenAIProvider.parse(data: data)
        case 401:
            throw AIError.authenticationFailed
        case 429:
            throw AIError.rateLimited
        default:
            let body = String(data: data, encoding: .utf8) ?? "<unreadable body>"
            throw AIError.sendFailed(message: "HTTP \(httpResponse.statusCode): \(body)")
        }
    }

    // MARK: - Request builder (internal for tests)

    static func buildRequest(
        endpoint: URL,
        apiKey: String,
        referer: String,
        title: String,
        messages: [ChatMessage],
        config: AIConfig
    ) throws -> URLRequest {
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.setValue(referer, forHTTPHeaderField: "HTTP-Referer")
        request.setValue(title, forHTTPHeaderField: "X-Title")

        var allMessages = messages
        if let systemPrompt = config.systemPrompt {
            allMessages.insert(ChatMessage(role: .system, content: systemPrompt), at: 0)
        }

        let body: [String: Any] = [
            "model": config.model,
            "max_tokens": config.maxTokens,
            "temperature": config.temperature,
            "messages": allMessages.map { msg -> [String: Any] in
                ["role": msg.role.rawValue, "content": serializeParts(msg.parts)]
            },
        ]

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            throw AIError.sendFailed(message: "Could not serialize request body: \(error.localizedDescription)")
        }
        return request
    }

    // MARK: - Private: part serialization

    private static func serializeParts(_ parts: [ChatMessage.Part]) -> Any {
        if parts.count == 1, case .text(let s) = parts[0] {
            return s
        }
        return parts.map { part -> [String: Any] in
            switch part {
            case .text(let s):
                return ["type": "text", "text": s]
            case .image(let jpegData, let mimeType):
                let dataURL = "data:\(mimeType);base64,\(jpegData.base64EncodedString())"
                return [
                    "type": "image_url",
                    "image_url": ["url": dataURL] as [String: Any],
                ]
            }
        }
    }
}

File: windows/src/KosmoNotes.Providers/OpenRouterProvider.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using KosmoNotes.Core.Models;
using KosmoNotes.Providers.Internal;

namespace KosmoNotes.Providers;

/// <summary>
/// <see cref="IAIProvider"/> for OpenRouter
/// (<c>POST https://openrouter.ai/api/v1/chat/completions</c>).
/// </summary>
/// <remarks>
/// OpenRouter is OpenAI-compatible — same body shape, same response parser.
/// Two notable differences:
/// <list type="bullet">
///   <item>Model identifier is <c>vendor/model</c> instead of a bare model name.</item>
///   <item>Adds <c>HTTP-Referer</c> + <c>X-Title</c> attribution headers; without
///         them OpenRouter rate-limits more aggressively.</item>
/// </list>
/// </remarks>
public sealed class OpenRouterProvider : IAIProvider
{
    /// <summary>The OpenRouter chat-completions endpoint.</summary>
    public static readonly Uri DefaultEndpoint =
        new("https://openrouter.ai/api/v1/chat/completions");

    /// <summary>Default <c>HTTP-Referer</c> header value.</summary>
    public const string DefaultReferer = "https://kosmonotes.studio";

    /// <summary>Default <c>X-Title</c> header value.</summary>
    public const string DefaultTitle = "Jarvis Note";

    private readonly string _apiKey;
    private readonly Uri _endpoint;
    private readonly string _referer;
    private readonly string _title;
    private readonly HttpClient _httpClient;

    /// <summary>Create a provider using the given API key.</summary>
    /// <param name="apiKey">OpenRouter API key (sent as <c>Authorization: Bearer …</c>).</param>
    /// <param name="endpoint">Optional endpoint override.</param>
    /// <param name="referer">Optional <c>HTTP-Referer</c> override.</param>
    /// <param name="title">Optional <c>X-Title</c> override.</param>
    /// <param name="httpClient">Optional client for tests.</param>
    public OpenRouterProvider(
        string apiKey,
        Uri? endpoint = null,
        string referer = DefaultReferer,
        string title = DefaultTitle,
        HttpClient? httpClient = null)
    {
        _apiKey = apiKey;
        _endpoint = endpoint ?? DefaultEndpoint;
        _referer = referer;
        _title = title;
        _httpClient = httpClient ?? HttpDefaults.Default;
    }

    /// <inheritdoc />
    public async Task<string> ChatAsync(
        IReadOnlyList<ChatMessage> messages,
        AIConfig config,
        CancellationToken ct = default)
    {
        using var request = BuildRequest(_endpoint, _apiKey, _referer, _title, messages, config);

        HttpResponseMessage response;
        try
        {
            response = await _httpClient.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (HttpRequestException ex)
        {
            throw new SendFailedException(ex.Message, ex);
        }

        try
        {
            await HttpStatus.ThrowIfNotSuccessAsync(response, ct).ConfigureAwait(false);
            byte[] body = await response.Content.ReadAsByteArrayAsync(ct).ConfigureAwait(false);
            // OpenRouter shares OpenAI's response shape — reuse its parser.
            return OpenAIProvider.Parse(body);
        }
        finally
        {
            response.Dispose();
        }
    }

    internal static HttpRequestMessage BuildRequest(
        Uri endpoint,
        string apiKey,
        string referer,
        string title,
        IReadOnlyList<ChatMessage> messages,
        AIConfig config)
    {
        var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
        request.Headers.TryAddWithoutValidation("HTTP-Referer", referer);
        request.Headers.TryAddWithoutValidation("X-Title", title);

        string json = OpenAIProvider.BuildBodyJson(messages, config);
        request.Content = new StringContent(json, Encoding.UTF8);
        request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
        return request;
    }
}
